### PR TITLE
nixos-test-driver: use a unique tmpdir for VDE sockets

### PR DESCRIPTION
**Pull request by delroth**
###### Description of changes

When running two nixos-test-driver instances that use VLANs on the same machine, the tmpdir used for vde_switch sockets overlap. This leads to broken virtual networking in at least one of the two sides, usually with a zombie vde_switch process.

There are other sources of conflicts/race conditions between nixos-test-driver instances (e.g. vm-state-* directories) but this one seems like the easiest to hit by far in my testing.

In theory this shouldn't impact runs on Hydra or via nix-build tests/... since they should be running in a sandbox with their own tmpdir. In practice, I'm seeing on Hydra very similar symptoms to the failures I've observed locally when this race condition is hit, and I'm not completely sure how to explain it just yet.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/lib/test-driver/test_driver/vlan.py [hunk 1]
```diff
@@ -24,7 +26,10 @@ def __repr__(self) -> str:
 
     def __init__(self, nr: int, tmp_dir: Path):
         self.nr = nr
-        self.socket_dir = tmp_dir / f"vde{self.nr}.ctl"
+        self.socket_tmp = tempfile.TemporaryDirectory(
+            prefix=f"vde{self.nr}.ctl", dir=tmp_dir
+        )
+        self.socket_dir = Path(self.socket_tmp.name)
```
Comment (tfc) [inline nixos/lib/test-driver/test_driver/vlan.py:32]: I would argue that teh variable `tmp_dir` should be globally set to distinct paths between multiple tests that are executed in parallel outside the sandbox.

This way we would still have all the temp files together per test, which is changed by this patch.

**Comment by delroth** [comment]
> This should not happen at all in the nix sandbox, you're right.
> 
> Plz help me understand your usecase... are you building the driver of multiple tests and execute them in parallel outside the sandbox, by running the built driver script in your local terminal?

Originally I found this issue when trying to debug some test flakiness on Hydra (for example: [vsftpd](https://hydra.nixos.org/log/9ap6bbxfzd6c8358v9wnrvyzzd35shc8-vm-test-run-vsftpd.drv)) -- as mentioned, I don't understand why it happens there just yet, but the symptoms are similar. I'm fairly convinced that this is not a broken test script but a driver/qemu/vde issue at this point.

While trying to repro locally I ended up encountering this issue. Running the test driver outside of the sandbox should be something that just works out of the box imo, I do it fairly often and it's supposed to be a [supported use case for development](https://nix.dev/tutorials/integration-testing-using-virtual-machines#developing-and-debugging-tests) (via the interactive runner target).

File: nixos/lib/test-driver/test_driver/vlan.py (hunk 1 above)
Comment (delroth) [reply nixos/lib/test-driver/test_driver/vlan.py:32]: Agreed. Question: do you care about the logic currently implemented by get_tmp_dir in driver.py? tempfile.gettempdir returning something that doesn't exist seems like a condition that most software wouldn't try to support and fix, and it would simplify things a lot if I could just rip that whole function off and replace with tempfile.TemporaryDirectory :)

Comment (delroth) [reply nixos/lib/test-driver/test_driver/vlan.py:32]: (I'm mostly wondering if you know of a common situation where that logic of `mkdir $TMPDIR` is actually useful -- it shouldn't be exercised in the Nix sandbox where /build is already there, and I can't imagine it happens on non-broken systems?)

**Comment by tfc** [comment]
> Originally I found this issue when trying to debug some test flakiness on Hydra (for example: [vsftpd](https://hydra.nixos.org/log/9ap6bbxfzd6c8358v9wnrvyzzd35shc8-vm-test-run-vsftpd.drv)) -- as mentioned, I don't understand why it happens there just yet, but the symptoms are similar. I'm fairly convinced that this is not a broken test script but a driver/qemu/vde issue at this point.

Sounds sane. I don't however consider this PR's changes a fix for those but completely orthogonal to those problems.

> While trying to repro locally I ended up encountering this issue. Running the test driver outside of the sandbox should be something that just works out of the box imo, I do it fairly often and it's supposed to be a [supported use case for development](https://nix.dev/tutorials/integration-testing-using-virtual-machines#developing-and-debugging-tests) (via the interactive runner target).

Yes, running the test driver outside of the sandbox interactively *is* a supported use case.
However, if you run multiple of them in parallel and all of them outside the sandbox (which is new to me as a use case because one would work interactively at one test at a time I guess), they should use completely distinct TMP dirs, not only for the vde.ctl sockets.
And that is already supported: You can always set different TMP dirs via an environment variable.
If many people end up needing this, i would suggest to extend the driver script with the functionality of randomizing its TMP dir (even outside the python code. the bash script which calls the python code could do that). If you're the only one who needs this, and only from time to time, then i don't consider it worthy of implementing such a special case.

What do you think?

File: nixos/lib/test-driver/test_driver/vlan.py (hunk 1 above)
Comment (tfc) [reply nixos/lib/test-driver/test_driver/vlan.py:32]: Sorry I have read what you wrote multiple times and I am a little bit confused. Can you please rephrase what you suggest to do?

Comment (delroth) [reply nixos/lib/test-driver/test_driver/vlan.py:32]: Let me rephrase: get_tmp_dir in driver.py right now doesn't make sense to me.

tempfile.gettempdir() is documented by Python to only return a *directory* that is *user writable*. If it can't find one, it returns cwd.

get_tmp_dir calls tempfile.gettempdir(), tries to mkdir the result, then checks that the result is a directory and that it's user writable. That's redundant with what tempfile.gettempdir() does.

Have I missed something that makes get_tmp_dir useful, or is ripping it out while refactoring fine?

**Comment by delroth** [comment]
> If many people end up needing this, i would suggest to extend the driver script with the functionality of randomizing its TMP dir (even outside the python code. the bash script which calls the python code could do that).

That's not the usual behavior of software that stores state in the tmpdir though. I don't think anybody would expect any program's temporary state to overwrite itself if you run two instances at once. POSIX has mktemp for a reason...

Is there any reason why fixing that working around the issue in a wrapper script is better than not having the problem in the first place?

File: nixos/lib/test-driver/test_driver/vlan.py (hunk 1 above)
Comment (tfc) [reply nixos/lib/test-driver/test_driver/vlan.py:32]: I understand where you're coming from.

The whole architecture is mostly inherited from the old implementation in Perl. From what it does, the function/dir could also be called "state dir" (which it is actually called in other parts of the code) because it's function is to collect in one folder structure what piles up during a test (socket files, host-guest shared dir, screenshots, ...).
This is controlled via the `TMPDIR` environment variable, which also bash scripts from within modules/virtualisation/qemu-vm.nix expect (i guess that predates even the integration test driver).
Also, this variable is documented in the test driver documentation. People will expect the driver to work in a way that they can set this variable and then all the test state shows up in there.
So we cannot just randomize it here.

From a UX perspective, it is very convenient (and i would argue that most people would expect it like this) that the test state shows up right in front of you.
If you're running multiple tests from the same folder outside the sandbox, i would still argue that you can give each invocation different TMPDIR env vars on their way.
